Readme.md
# UP 中国区使用方法

[UP](https://github.com/apex/up) 是一个快速创建、部署 serverless 的服务。使用文档参考 [https://up.docs.apex.sh/](https://up.docs.apex.sh/)。

因为中国大陆 AWS 服务与其它地区的 API 有些区别，所以 UP 暂时不能在中国区使用。为了解决这个问题，在 UP v1.2.2 代码基础上做了些改动。目前达到**可用**状态。

## 改动部分
1. 增加了 cn-north-1 / cn-north-west-1 区域，分别代表 AWS 北京和 AWS 宁夏区域；（文件：./platform/aws/regions/regions.go）
2. 注释了 environment variables 相关代码，中国区暂不支持；（文件： ./platform/lambda/lambda.go）
3. 为 API Gateway 和 Lambda 增加了 .cn 的 endpoint；参考 [AWS 区域和终端节点](http://docs.amazonaws.cn/general/latest/gr/rande.html)。（文件： ./platform/lambda/lambda.go）
4. Binary 名称改为 upcn。（文件： install.sh 和 .goreleaser.yml）

## 安装
```
$ curl -sf https://raw.githubusercontent.com/leplay/upcn/master/install.sh | sh
```

## 使用方法
1. 重要！如果要在中国区使用 API Gateway 并且要使用公开 url 访问，需要去 AWS 支持中心创建案例申请开通；
2. 创建 up.json 要指定区域，并且将 cors 设置为 null，不要设置 environment 属性。示例：
```
{
  "name": "test",
  "profile": "up-cn",
  "regions": ["cn-north-1"],
  "cors": null
}
```
3. 在脚本目录里执行：
```
export AWS_REGION=cn-north-1
```
4. upcn deploy 之后，需要手动去 [AWS Lambda 后台](https://console.amazonaws.cn/lambda/home?region=cn-north-1) 添加一下 API Gateway 触发器。


History.md

v861.2.2 / 2019-05-23
==================

  * AWS China Support

v1.2.2 / 2019-05-23
==================

  * add ./vendor to excluded directories by default
  * docs
  * docs: adjust multi-region mentions for now
  * docs
  * Release v1.2.0
  * add 60% coupon

v1.2.0 / 2019-04-20
===================

  * add 60% coupon

v1.2.0-pro / 2019-04-05
=======================

  * add regional DNS with latency-based routing
  * add `--region` flag for every command to override region id

v1.1.3-pro / 2019-04-02
=======================

  * Rebase

v1.1.3 / 2019-04-02
===================

  * update tj/aws dependency for duplicate logs fix

v1.1.2-pro / 2019-03-29
=======================

  * Rebase

v1.1.2 / 2019-03-29
===================

  * fix: update tj/aws dependency for ThrottlingException logs exception

v1.1.1-pro / 2019-03-23
=======================

  * update warming function to nodejs8.10 to prevent EOL warning from AWS

v1.1.0-pro / 2019-03-04
===================

  * add file based environment variables, removing the size restrictions

v1.0.0-pro / 2019-02-26
===================

  * add regional endpoint support

v0.9.1-pro / 2019-01-21
=======================

  * add sorting of env vars. Closes #750

v0.9.0-pro / 2018-12-13
=======================

  * add Lambda layer support. Closes #743

v0.8.1-pro / 2018-12-11
=======================

  * improve `up env export` performance, no longer linear time
  * fix `up deploys` error when the stage is not deployed. Closes #716
  * Rebase

v0.8.1 / 2018-12-11
===================

  * update tj/aws for bug preventing all logs from being returned. Closes #733
  * add deploy --no-build flag for skipping build hooks. Closes #730
  * Release v0.8.0-pro

v0.8.0-pro / 2018-12-06
=======================

  * improve `up env export` performance, no longer linear time
  * fix `up deploys` error when the stage is not deployed. Closes #716

v0.8.0 / 2018-12-04
===================

  * add endpoint URL to the deployment output
  * add deploy stage to the deployment output
  * add msg about Up Pro
  * remove "not info" log example. Closes #724
  * fix typo in deploy example. Closes #718
  * fix: use crystallang/crystal for Crystal builds (#713)

v0.7.8-pro / 2018-09-24
=======================

  * add `up env export` command for exposing env vars to shell scripts

v0.7.7-pro / 2018-09-17
=======================

  * Rebase

v0.7.7 / 2018-09-17
===================

  * update go-update dependency for copy regression

v0.7.5-pro / 2018-09-17
=======================

  * Rebase

v0.7.5 / 2018-09-17
===================

  * update go-update dependency for rename() to copy replacement

v0.7.4-pro / 2018-09-16
=======================

  * add baked in env vars from SSM, env vars are no longer loaded at runtime. Closes #547

v0.7.4 / 2018-09-16
===================

  * add request id to proxy errors
  * fix missing lambda configurations costs. (#703)
  * fix order relay errors so timeouts are returned first (leads to better error messages)

v0.7.3-pro / 2018-08-08
=======================

  * Rebase

v0.7.3 / 2018-08-08
===================

  * fix crash recovery in lambda, bug was introduced in v0.7.0
  * update cors middleware for security when using allow-origin * and allow-credentials

v0.7.2-pro / 2018-07-23
=======================

  * Rebase

v0.7.2 / 2018-07-23
===================

  * add vpc stage override support. Closes #689

v0.7.1-pro / 2018-07-12
=======================

  * Rebase

v0.7.1 / 2018-07-12
===================

  * fix initial IAM role creation waiting due to error response change

v0.7.0-pro / 2018-07-11
=======================

  * Rebase

v0.7.0 / 2018-07-11
===================

  * add in-flight request timeouts.
  * remove retries
  * refactor crash recovery to be more robust

v0.6.8-pro / 2018-06-07
=======================

  * Rebase

v0.6.8 / 2018-06-07
===================

  * fix multiple set-cookie API Gateway limitation for real (previous had a bug)

v0.6.7-pro / 2018-06-07
=======================

  * fix s3 acceleration update with existing S3 buckets
  * Rebase

v0.6.7 / 2018-06-07
===================

  * add striping of @owner/repo@ portion of Lerna tags. Closes #670
  * fix multiple set-cookies API Gateway limitation with casing hack
  * fix deployment with empty Git repo
  * update AWS SDK versions for assuming roles. (#668)

v0.6.6-pro / 2018-05-24
=======================

  * Rebase

v0.6.6 / 2018-05-24
===================

  * add vpc support. Closes #281
  * fix Crystal build on Linux: PWD => pwd (#664)

v0.6.5-pro / 2018-05-16
=======================

  * Rebase

v0.6.5 / 2018-05-16
===================

  * add hidden disable stats command (#659)
  * add X-Context header field. Closes #657
  * fix CORS header fields from being clobbered by error pages. Closes #661

v0.6.4-pro / 2018-05-09
=======================

  * add deployment size to `up deploys` output
  * add asterisk to denote current version in `up deploys` due to rollbacks
  * Rebase

v0.6.4 / 2018-05-09
===================

  * add support for customizing the Lambda function IAM role policy. Closes #539
  * add support for specifying dns zone, and disabling it. Closes #536
  * add support for updating the role policy upon deploy
  * change default prune retention to 30 versions

v0.6.3-pro / 2018-05-02
=======================

  * add deployment size to `up deploys` output
  * add asterisk to denote current version in `up deploys` due to rollbacks
  * Rebase

v0.6.3 / 2018-05-02
===================

  * add `--stage` flag to `up build`
  * add `--stage` flag to `up run`
  * change logs, metrics, and url commands to use `-s` flag for stage. Closes #371 (BREAKING)

v0.6.2-pro / 2018-04-25
=======================

  * Rebase

v0.6.2 / 2018-04-25
===================

  * add up prune `--stage` flag. Closes #647
  * add `up` to ignore whitelist by default
  * remove retries on 5xx. Closes #485

platform/lambda/stack/resources/resources_test.go
package resources

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/apex/up"
	"github.com/apex/up/config"
)

// keys returns keys from a map.
func keys(m Map) (v []string) {
	for k := range m {
		v = append(v, k)
	}
	return
}

// getResource returns resource by name.
func getResource(c *Config, name string) Map {
	tmpl := New(c)
	r := tmpl["Resources"].(Map)

	v, ok := r[name].(Map)
	if !ok {
		k := strings.Join(keys(r), "\n  - ")
		panic(fmt.Sprintf("resource %q does not exist in:\n\n  - %s", name, k))
	}

	return v
}

// dump a resource to stdout.
func dump(c *Config, name string) {
	r := getResource(c, name)
	{
		enc := json.NewEncoder(os.Stdout)
		enc.SetIndent("", "  ")
		enc.Encode(r)
	}
}

func Example_api() {
	c := &Config{
		Config: &up.Config{
			Name: "polls",
		},
	}

	dump(c, "Api")
	// Output:
	// {
	//   "Properties": {
	//     "BinaryMediaTypes": [
	//       "*/*"
	//     ],
	//     "Description": "Managed by Up.",
	//     "Name": {
	//       "Ref": "Name"
	//     }
	//   },
	//   "Type": "AWS::ApiGateway::RestApi"
	// }
}

func Example_apiRootMethod() {
	c := &Config{
		Config: &up.Config{
			Name: "polls",
		},
	}

	dump(c, "ApiRootMethod")
	// Output:
	// {
	//   "Properties": {
	//     "AuthorizationType": "NONE",
	//     "HttpMethod": "ANY",
	//     "Integration": {
	//       "IntegrationHttpMethod": "POST",
	//       "Type": "AWS_PROXY",
	//       "Uri": {
	//         "Fn::Join": [
	//           "",
	//           [
	//             "arn:aws:apigateway:",
	//             {
	//               "Ref": "AWS::Region"
	//             },
	//             ":lambda:path/2015-03-31/functions/",
	//             {
	//               "Fn::Join": [
	//                 ":",
	//                 [
	//                   "arn",
	//                   "aws",
	//                   "lambda",
	//                   {
	//                     "Ref": "AWS::Region"
	//                   },
	//                   {
	//                     "Ref": "AWS::AccountId"
	//                   },
	//                   "function",
	//                   {
	//                     "Fn::Join": [
	//                       ":",
	//                       [
	//                         {
	//                           "Ref": "FunctionName"
	//                         },
	//                         "${stageVariables.qualifier}"
	//                       ]
	//                     ]
	//                   }
	//                 ]
	//               ]
	//             },
	//             "/invocations"
	//           ]
	//         ]
	//       }
	//     },
	//     "ResourceId": {
	//       "Fn::GetAtt": [
	//         "Api",
	//         "RootResourceId"
	//       ]
	//     },
	//     "RestApiId": {
	//       "Ref": "Api"
	//     }
	//   },
	//   "Type": "AWS::ApiGateway::Method"
	// }
}

func Example_apiProxyResource() {
	c := &Config{
		Config: &up.Config{
			Name: "polls",
		},
	}

	dump(c, "ApiProxyResource")
	// Output:
	// {
	//   "Properties": {
	//     "ParentId": {
	//       "Fn::GetAtt": [
	//         "Api",
	//         "RootResourceId"
	//       ]
	//     },
	//     "PathPart": "{proxy+}",
	//     "RestApiId": {
	//       "Ref": "Api"
	//     }
	//   },
	//   "Type": "AWS::ApiGateway::Resource"
	// }
}

func Example_apiProxyMethod() {
	c := &Config{
		Config: &up.Config{
			Name: "polls",
		},
	}

	dump(c, "ApiProxyMethod")
	// Output:
	// {
	//   "Properties": {
	//     "AuthorizationType": "NONE",
	//     "HttpMethod": "ANY",
	//     "Integration": {
	//       "IntegrationHttpMethod": "POST",
	//       "Type": "AWS_PROXY",
	//       "Uri": {
	//         "Fn::Join": [
	//           "",
	//           [
	//             "arn:aws:apigateway:",
	//             {
	//               "Ref": "AWS::Region"
	//             },
	//             ":lambda:path/2015-03-31/functions/",
	//             {
	//               "Fn::Join": [
	//                 ":",
	//                 [
	//                   "arn",
	//                   "aws",
	//                   "lambda",
	//                   {
	//                     "Ref": "AWS::Region"
	//                   },
	//                   {
	//                     "Ref": "AWS::AccountId"
	//                   },
	//                   "function",
	//                   {
	//                     "Fn::Join": [
	//                       ":",
	//                       [
	//                         {
	//                           "Ref": "FunctionName"
	//                         },
	//                         "${stageVariables.qualifier}"
	//                       ]
	//                     ]
	//                   }
	//                 ]
	//               ]
	//             },
	//             "/invocations"
	//           ]
	//         ]
	//       }
	//     },
	//     "ResourceId": {
	//       "Ref": "ApiProxyResource"
	//     },
	//     "RestApiId": {
	//       "Ref": "Api"
	//     }
	//   },
	//   "Type": "AWS::ApiGateway::Method"
	// }
}

func Example_stageAlias() {
	c := &Config{
		Config: &up.Config{
			Name: "polls",
			Stages: config.Stages{
				"production": &config.Stage{
					Name: "production",
				},
			},
		},
		Versions: Versions{
			"production": "15",
		},
	}

	dump(c, "ApiFunctionAliasProduction")
	// Output:
	// {
	//   "Properties": {
	//     "Description": "Managed by Up.",
	//     "FunctionName": {
	//       "Ref": "FunctionName"
	//     },
	//     "FunctionVersion": "15",
	//     "Name": "production"
	//   },
	//   "Type": "AWS::Lambda::Alias"
	// }
}

func Example_stagePermission() {
	c := &Config{
		Config: &up.Config{
			Name: "polls",
			Stages: config.Stages{
				"production": &config.Stage{
					Name: "production",
				},
			},
		},
		Versions: Versions{
			"production": "15",
		},
	}

	dump(c, "ApiLambdaPermissionProduction")
	// Output:
	// {
	//   "DependsOn": "ApiFunctionAliasProduction",
	//   "Properties": {
	//     "Action": "lambda:invokeFunction",
	//     "FunctionName": {
	//       "Fn::Join": [
	//         ":",
	//         [
	//           "arn",
	//           "aws",
	//           "lambda",
	//           {
	//             "Ref": "AWS::Region"
	//           },
	//           {
	//             "Ref": "AWS::AccountId"
	//           },
	//           "function",
	//           {
	//             "Fn::Join": [
	//               ":",
	//               [
	//                 {
	//                   "Ref": "FunctionName"
	//                 },
	//                 "production"
	//               ]
	//             ]
	//           }
	//         ]
	//       ]
	//     },
	//     "Principal": "apigateway.amazonaws.com.cn",
	//     "SourceArn": {
	//       "Fn::Join": [
	//         "",
	//         [
	//           "arn:aws:execute-api",
	//           ":",
	//           {
	//             "Ref": "AWS::Region"
	//           },
	//           ":",
	//           {
	//             "Ref": "AWS::AccountId"
	//           },
	//           ":",
	//           {
	//             "Ref": "Api"
	//           },
	//           "/*"
	//         ]
	//       ]
	//     }
	//   },
	//   "Type": "AWS::Lambda::Permission"
	// }
}

func Example_stageDeployment() {
	c := &Config{
		Config: &up.Config{
			Name: "polls",
			Stages: config.Stages{
				"production": &config.Stage{
					Name: "production",
				},
			},
		},
		Versions: Versions{
			"production": "15",
		},
	}

	dump(c, "ApiDeploymentProduction")
	// Output:
	// {
	//   "DependsOn": [
	//     "ApiRootMethod",
	//     "ApiProxyMethod",
	//     "ApiFunctionAliasProduction"
	//   ],
	//   "Properties": {
	//     "RestApiId": {
	//       "Ref": "Api"
	//     },
	//     "StageDescription": {
	//       "Variables": {
	//         "qualifier": "production"
	//       }
	//     },
	//     "StageName": "production"
	//   },
	//   "Type": "AWS::ApiGateway::Deployment"
	// }
}

func Example_stageDomain() {
	c := &Config{
		Config: &up.Config{
			Name: "polls",
			Stages: config.Stages{
				"production": &config.Stage{
					Name:   "production",
					Domain: "up-example.com",
					Cert:   "arn::something",
				},
			},
		},
		Versions: Versions{
			"production": "15",
		},
	}

	dump(c, "ApiDomainProduction")
	// Output:
	// 	{
	//   "Properties": {
	//     "CertificateArn": "arn::something",
	//     "DomainName": "up-example.com"
	//   },
	//   "Type": "AWS::ApiGateway::DomainName"
	// }
}

func Example_stagePathMapping() {
	c := &Config{
		Config: &up.Config{
			Name: "polls",
			Stages: config.Stages{
				"production": &config.Stage{
					Name:   "production",
					Domain: "up-example.com",
				},
			},
		},
		Versions: Versions{
			"production": "15",
		},
	}

	dump(c, "ApiDomainProductionPathMapping")
	// Output:
	// {
	//   "DependsOn": [
	//     "ApiDeploymentProduction",
	//     "ApiDomainProduction"
	//   ],
	//   "Properties": {
	//     "BasePath": "",
	//     "DomainName": "up-example.com",
	//     "RestApiId": {
	//       "Ref": "Api"
	//     },
	//     "Stage": "production"
	//   },
	//   "Type": "AWS::ApiGateway::BasePathMapping"
	// }
}

func Example_stageDNSZone() {
	c := &Config{
		Config: &up.Config{
			Name: "polls",
			Stages: config.Stages{
				"production": &config.Stage{
					Name:   "production",
					Domain: "up-example.com",
				},
			},
		},
		Versions: Versions{
			"production": "15",
		},
	}

	dump(c, "DnsZoneUpExampleCom")
	// Output:
	// {
	//   "Properties": {
	//     "Name": "up-example.com"
	//   },
	//   "Type": "AWS::Route53::HostedZone"
	// }
}

func Example_stageDNSZoneRecord() {
	c := &Config{
		Config: &up.Config{
			Name: "polls",
			Stages: config.Stages{
				"production": &config.Stage{
					Name:   "production",
					Domain: "up-example.com",
				},
			},
		},
		Versions: Versions{
			"production": "15",
		},
	}

	dump(c, "DnsZoneUpExampleComRecordUpExampleCom")
	// Output:
	// {
	//   "Properties": {
	//     "AliasTarget": {
	//       "DNSName": {
	//         "Fn::GetAtt": [
	//           "ApiDomainProduction",
	//           "DistributionDomainName"
	//         ]
	//       },
	//       "HostedZoneId": "Z2FDTNDATAQYW2"
	//     },
	//     "Comment": "Managed by Up.",
	//     "HostedZoneId": {
	//       "Ref": "DnsZoneUpExampleCom"
	//     },
	//     "Name": "up-example.com",
	//     "Type": "A"
	//   },
	//   "Type": "AWS::Route53::RecordSet"
	// }
}

func Example_dnsZone() {
	c := &Config{
		Config: &up.Config{
			Name: "polls",
			DNS: config.DNS{
				Zones: []*config.Zone{
					{
						Name: "up-example.com",
					},
				},
			},
		},
	}

	dump(c, "DnsZoneUpExampleCom")
	// Output:
	// {
	//   "Properties": {
	//     "Name": "up-example.com"
	//   },
	//   "Type": "AWS::Route53::HostedZone"
	// }
}

func Example_dnsZoneRecord() {
	c := &Config{
		Config: &up.Config{
			Name: "polls",
			DNS: config.DNS{
				Zones: []*config.Zone{
					{
						Name: "up-example.com",
						Records: []*config.Record{
							{
								Name:  "blog.up-example.com",
								Type:  "CNAME",
								TTL:   600,
								Value: []string{"example.medium.com"},
							},
						},
					},
				},
			},
		},
	}

	dump(c, "DnsZoneUpExampleComRecordBlogUpExampleComCNAME")
	// Output:
	// {
	//   "Properties": {
	//     "Comment": "Managed by Up.",
	//     "HostedZoneId": {
	//       "Ref": "DnsZoneUpExampleCom"
	//     },
	//     "Name": "blog.up-example.com",
	//     "ResourceRecords": [
	//       "example.medium.com"
	//     ],
	//     "TTL": "600",
	//     "Type": "CNAME"
	//   },
	//   "Type": "AWS::Route53::RecordSet"
	// }
}


platform/aws/regions/regions.go
// Package regions provides AWS region utilities.
package regions

import (
	"fmt"
	"path/filepath"
)

// hostedZoneIDs is a set of hosted zone ids for API Gateway.
var hostedZoneIDs = map[string]string{
	"us-east-2":      "ZOJJZC49E0EPZ",
	"us-east-1":      "Z1UJRXOUMOOFQ8",
	"us-west-1":      "Z2MUQ32089INYE",
	"us-west-2":      "Z2OJLYMUO9EFXC",
	"ap-south-1":     "Z3VO1THU9YC4UR",
	"ap-northeast-3": "Z2YQB5RD63NC85",
	"ap-northeast-2": "Z20JF4UZKIW1U8",
	"ap-southeast-1": "ZL327KTPIQFUL",
	"ap-southeast-2": "Z2RPCDW04V8134",
	"ap-northeast-1": "Z1YSHQZHG15GKL",
	"ca-central-1":   "Z19DQILCV0OWEC",
	"eu-central-1":   "Z1U9ULNL0V5AJ3",
	"eu-west-1":      "ZLY8HYME6SFDD",
	"eu-west-2":      "ZJ5UAJN8Y3Z2Q",
	"eu-west-3":      "Z3KY65QIEKYHQQ",
	"eu-north-1":     "Z2YB950C88HT6D",
	"sa-east-1":      "ZCMLWB8V5SYIT",
}

// IDs of regions.
var IDs = []string{
	"us-east-2",
	"us-east-1",
	"us-west-1",
	"us-west-2",
	"ap-south-1",
	"ap-northeast-2",
	"ap-southeast-1",
	"ap-southeast-2",
	"ap-northeast-1",
	"ca-central-1",
	"eu-central-1",
	"eu-west-1",
	"eu-west-2",
	"eu-west-3",
	"eu-north-1",
	"sa-east-1",
	"cn-north-1",
	"cn-northwest-1",
}

// Names of regions.
var Names = []string{
	"US East (Ohio)",
	"US East (N. Virginia)",
	"US West (N. California)",
	"US West (Oregon)",
	"Asia Pacific (Mumbai)",
	"Asia Pacific (Seoul)",
	"Asia Pacific (Singapore)",
	"Asia Pacific (Sydney)",
	"Asia Pacific (Tokyo)",
	"Canada (Central)",
	"EU (Frankfurt)",
	"EU (Ireland)",
	"EU (London)",
	"EU (Paris)",
	"EU (Stockholm)",
	"South America (São Paulo)",
	"China (Beijing)",
	"China (Ningxia)",
}

// Match returns regions matching the pattern(s) provided. Patterns
// which are not "expanded" are returned as-is.
func Match(regions []string) (v []string) {
	for _, pattern := range regions {
		matched := false

		for _, id := range IDs {
			if ok, _ := filepath.Match(pattern, id); ok {
				v = append(v, id)
				matched = true
			}
		}

		if !matched {
			v = append(v, pattern)
		}
	}

	return
}

// GetIdByName returns a region id by name.
func GetIdByName(name string) string {
	for i, n := range Names {
		if n == name {
			return IDs[i]
		}
	}
	return ""
}

// GetHostedZoneID returns a hosted zone id by region.
func GetHostedZoneID(region string) string {
	id, ok := hostedZoneIDs[region]
	if !ok {
		panic(fmt.Sprintf("region %q is not yet supported", region))
	}
	return id
}


platform/lambda/stack/resources/resources.go
package resources

import (
	"fmt"
	"strconv"

	"github.com/apex/up"
	"github.com/apex/up/config"
	"github.com/apex/up/internal/util"
	"github.com/aws/aws-sdk-go/service/route53"
)

// Map type.
type Map map[string]interface{}

// Versions is a map of stage to lambda function version.
type Versions map[string]string

// Config for the resource template.
type Config struct {
	// Zones already present in route53. This is used to
	// ensure that existing zones previously set up, or
	// automatically configured when purchasing a domain
	// are not duplicated.
	Zones []*route53.HostedZone

	// Versions map used to maintain the correct lambda
	// function aliases when updating a stack.
	Versions Versions

	*up.Config
}

// New template.
func New(c *Config) map[string]interface{} {
	return Map{
		"AWSTemplateFormatVersion": "2010-09-09",
		"Parameters":               parameters(c),
		"Outputs":                  outputs(c),
		"Resources":                resources(c),
	}
}

// ref of id.
func ref(id string) Map {
	return Map{
		"Ref": id,
	}
}

// get value from named ref.
func get(name, value string) Map {
	return Map{
		"Fn::GetAtt": []string{
			name,
			value,
		},
	}
}

// join strings with delim.
func join(delim string, s ...interface{}) Map {
	return Map{
		"Fn::Join": []interface{}{
			delim,
			s,
		},
	}
}

// stageVariable by name.
func stageVariable(name string) string {
	return fmt.Sprintf("${stageVariables.%s}", name)
}

// lambda ARN for function name.
func lambdaArn(name string) Map {
	return join(":", "arn", "aws", "lambda", ref("AWS::Region"), ref("AWS::AccountId"), "function", ref(name))
}

// lambda ARN for function name with qualifier.
func lambdaArnQualifier(name, qualifier string) Map {
	return join(":", "arn", "aws", "lambda", ref("AWS::Region"), ref("AWS::AccountId"), "function", join(":", ref(name), qualifier))
}

// getZone returns a zone by domain or nil.
func getZone(c *Config, domain string) *route53.HostedZone {
	for _, z := range c.Zones {
		if *z.Name == domain+"." {
			return z
		}
	}
	return nil
}

// dnsZone returns the ref to a new zone, or id to an existing zone.
func dnsZone(c *Config, m Map, domain string) interface{} {
	// already exists
	if z := getZone(c, domain); z != nil {
		return *z.Id
	}

	id := util.Camelcase("dns_zone_%s", domain)

	// already registered for creation
	if m[id] != nil {
		return ref(id)
	}

	// new zone
	m[id] = Map{
		"Type": "AWS::Route53::HostedZone",
		"Properties": Map{
			"Name": domain,
		},
	}

	return ref(id)
}

// api sets up the app resources.
func api(c *Config, m Map) {
	m["Api"] = Map{
		"Type": "AWS::ApiGateway::RestApi",
		"Properties": Map{
			"Name":        ref("Name"),
			"Description": util.ManagedByUp(c.Description),
			"BinaryMediaTypes": []string{
				"*/*",
			},
		},
	}

	integration := Map{
		"Type":                  "AWS_PROXY",
		"IntegrationHttpMethod": "POST",
		"Uri": join("",
			"arn:aws:apigateway:",
			ref("AWS::Region"),
			":lambda:path/2015-03-31/functions/",
			lambdaArnQualifier("FunctionName", stageVariable("qualifier")),
			"/invocations"),
	}

	m["ApiRootMethod"] = Map{
		"Type": "AWS::ApiGateway::Method",
		"Properties": Map{
			"RestApiId":         ref("Api"),
			"ResourceId":        get("Api", "RootResourceId"),
			"HttpMethod":        "ANY",
			"AuthorizationType": "NONE",
			"Integration":       integration,
		},
	}

	m["ApiProxyResource"] = Map{
		"Type": "AWS::ApiGateway::Resource",
		"Properties": Map{
			"RestApiId": ref("Api"),
			"ParentId":  get("Api", "RootResourceId"),
			"PathPart":  "{proxy+}",
		},
	}

	m["ApiProxyMethod"] = Map{
		"Type": "AWS::ApiGateway::Method",
		"Properties": Map{
			"RestApiId":         ref("Api"),
			"ResourceId":        ref("ApiProxyResource"),
			"HttpMethod":        "ANY",
			"AuthorizationType": "NONE",
			"Integration":       integration,
		},
	}

	stages(c, m)
}

// stages sets up the stage specific resources.
func stages(c *Config, m Map) {
	for _, s := range c.Stages.List() {
		if s.IsRemote() {
			stage(c, s, m)
		}
	}
}

// stage sets up the stage specific resources.
func stage(c *Config, s *config.Stage, m Map) {
	aliasID := stageAlias(c, s, m)
	deploymentID := stageDeployment(c, s, m, aliasID)
	stagePermissions(c, s, m, aliasID)
	stageDomain(c, s, m, deploymentID)
}

// stageAlias sets up the lambda alias and deployment and returns the alias id.
func stageAlias(c *Config, s *config.Stage, m Map) string {
	id := util.Camelcase("api_function_alias_%s", s.Name)
	version, ok := c.Versions[s.Name]

	if !ok {
		panic(fmt.Sprintf("stage %q is missing a function version mapping", s.Name))
	}

	m[id] = Map{
		"Type": "AWS::Lambda::Alias",
		"Properties": Map{
			"Name":            s.Name,
			"Description":     util.ManagedByUp(""),
			"FunctionName":    ref("FunctionName"),
			"FunctionVersion": version,
		},
	}

	return id
}

// stagePermissions sets up the lambda:invokeFunction permissions for API Gateway.
func stagePermissions(c *Config, s *config.Stage, m Map, aliasID string) {
	id := util.Camelcase("api_lambda_permission_%s", s.Name)

	m[id] = Map{
		"Type":      "AWS::Lambda::Permission",
		"DependsOn": aliasID,
		"Properties": Map{
			"Action":       "lambda:invokeFunction",
			"FunctionName": lambdaArnQualifier("FunctionName", s.Name),
			"Principal":    "apigateway.amazonaws.com.cn",
			"SourceArn": join("",
				"arn:aws:execute-api",
				":",
				ref("AWS::Region"),
				":",
				ref("AWS::AccountId"),
				":",
				ref("Api"),
				"/*"),
		},
	}
}

// stageDeployment sets up the API Gateway deployment.
func stageDeployment(c *Config, s *config.Stage, m Map, aliasID string) string {
	id := util.Camelcase("api_deployment_%s", s.Name)

	m[id] = Map{
		"Type":      "AWS::ApiGateway::Deployment",
		"DependsOn": []string{"ApiRootMethod", "ApiProxyMethod", aliasID},
		"Properties": Map{
			"RestApiId": ref("Api"),
			"StageName": s.Name,
			"StageDescription": Map{
				"Variables": Map{
					"qualifier": s.Name,
				},
			},
		},
	}

	return id
}

// stageDomain sets up a custom domain, dns record and path mapping.
func stageDomain(c *Config, s *config.Stage, m Map, deploymentID string) {
	if s.Domain == "" {
		return
	}

	id := util.Camelcase("api_domain_%s", s.Name)

	m[id] = Map{
		"Type": "AWS::ApiGateway::DomainName",
		"Properties": Map{
			"CertificateArn": s.Cert,
			"DomainName":     s.Domain,
		},
	}

	stagePathMapping(c, s, m, deploymentID, id)

	if s.Zone != false {
		stageDNSRecord(c, s, m, id)
	}
}

// stagePathMapping sets up the stage deployment mapping.
func stagePathMapping(c *Config, s *config.Stage, m Map, deploymentID, domainID string) {
	id := util.Camelcase("api_domain_%s_path_mapping", s.Name)

	m[id] = Map{
		"Type":      "AWS::ApiGateway::BasePathMapping",
		"DependsOn": []string{deploymentID, domainID},
		"Properties": Map{
			"DomainName": s.Domain,
			"BasePath":   util.BasePath(s.Path),
			"RestApiId":  ref("Api"),
			"Stage":      s.Name,
		},
	}
}

// stageDNSRecord sets up an ALIAS record and zone if necessary for a custom domain.
func stageDNSRecord(c *Config, s *config.Stage, m Map, domainID string) {
	id := util.Camelcase("dns_zone_%s_record_%s", util.Domain(s.Domain), s.Domain)
	zoneName := util.Domain(s.Domain)

	// explicit .zone was specified
	if s, ok := s.Zone.(string); ok {
		zoneName = s
	}

	zone := dnsZone(c, m, zoneName)

	m[id] = Map{
		"Type": "AWS::Route53::RecordSet",
		"Properties": Map{
			"Name":         s.Domain,
			"Type":         "A",
			"Comment":      util.ManagedByUp(""),
			"HostedZoneId": zone,
			"AliasTarget": Map{
				"DNSName":      get(domainID, "DistributionDomainName"),
				"HostedZoneId": "Z2FDTNDATAQYW2",
			},
		},
	}
}

// dns setups the the user-defined DNS zones and records.
func dns(c *Config, m Map) {
	for _, z := range c.DNS.Zones {
		zone := dnsZone(c, m, z.Name)

		for _, r := range z.Records {
			id := util.Camelcase("dns_zone_%s_record_%s_%s", z.Name, r.Name, r.Type)

			m[id] = Map{
				"Type": "AWS::Route53::RecordSet",
				"Properties": Map{
					"Name":            r.Name,
					"Type":            r.Type,
					"TTL":             strconv.Itoa(r.TTL),
					"ResourceRecords": r.Value,
					"HostedZoneId":    zone,
					"Comment":         util.ManagedByUp(""),
				},
			}
		}
	}
}

// resources of the stack.
func resources(c *Config) Map {
	m := Map{}
	api(c, m)
	dns(c, m)
	return m
}

// parameters of the stack.
func parameters(c *Config) Map {
	return Map{
		"Name": Map{
			"Description": "Name of application",
			"Type":        "String",
		},
		"FunctionName": Map{
			"Description": "Name of application function",
			"Type":        "String",
		},
	}
}

// outputs of the stack.
func outputs(c *Config) Map {
	return Map{
		"ApiName": Map{
			"Description": "API name",
			"Value":       ref("Name"),
		},
		"ApiFunctionName": Map{
			"Description": "API Lambda function name",
			"Value":       ref("FunctionName"),
		},
		"ApiFunctionArn": Map{
			"Description": "API Lambda function ARN",
			"Value":       lambdaArn("FunctionName"),
		},
	}
}
